Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py

```diff
@@ -0,0 +1,597 @@
+    for channelSetting in context.scene.meshsync_bake_channel_settings:
+        channelSetting.bakeChannelEnabled = context.scene.meshsync_bake_all_channels
+
+def msb_bakeAllGet(scene):
+    for channelSetting in scene.meshsync_bake_channel_settings:
+        if not channelSetting.bakeChannelEnabled:
+            return False
+    return True
+
+def msb_bakeAllSet(scene, newValue):
+    for channelSetting in scene.meshsync_bake_channel_settings:
+        channelSetting.bakeChannelEnabled = newValue
+
+@persistent
+def msb_setBakingDefaults(dummy):
+    context = bpy.context
+    context.scene.meshsync_bake_channel_settings.clear()
+    if len(context.scene.meshsync_bake_channel_settings) != len(BAKED_CHANNELS):
+        context.scene.meshsync_bake_channel_settings.clear()
+        for channel in BAKED_CHANNELS:
+            channelSettings = context.scene.meshsync_bake_channel_settings.add()
+            channelSettings.name = channel
+
+class MESHSYNC_PT_Baking(MESHSYNC_PT, bpy.types.Panel):
+    bl_label = "Baking"
+    bl_parent_id = "MESHSYNC_PT_Main"
+    bl_order = -10
+
+    def draw(self, context):
+        layout = self.layout
+        layout.use_property_split = True
+        layout.use_property_decorate = False
+
+        layout.prop(context.scene, "meshsync_bake_selection", expand=True)
+
+        box = layout# layout.box()
+        box.alignment = 'LEFT'
+        box.label(text="Material channels to bake")
+        box.prop(context.scene, "meshsync_bake_all_channels")
+        for channelSetting in context.scene.meshsync_bake_channel_settings:
+            row = box.row()
+            row.prop(channelSetting, "bakeChannelEnabled", text=channelSetting.name)
+
+        layout.operator("meshsync.bake_materials")
+        row = layout.row()
+        row.prop(context.scene, "meshsync_bakedTexturesPath")
+        row.operator("meshsync.choose_material_bake_folder", icon="FILE_FOLDER", text="")
+        layout.prop(context.scene, "meshsync_bakedTextureSize")
+        layout.operator("meshsync.revert_bake_materials")
+
+
+class MESHSYNC_OT_Bake(bpy.types.Operator):
+    bl_idname = "meshsync.bake_materials"
+    bl_label = "Bake to individual materials"
+    bl_description = "Bakes textures, creates material copies and assigns the baked materials for all materials that " \
+                     "cannot be exported without baking them to textures"
+
+    def isMaterialCopy(self, mat):
+        return ORIGINAL_MATERIAL in mat
+
+    def canMaterialBeBaked(self, mat):
+        return mat is not None and\
+            not self.isMaterialCopy(mat) and\
+            mat.use_nodes
+
+    def findMaterialOutputNode(self, mat):
+        node_tree = mat.node_tree
+
+        outputNode = None
+        for node in node_tree.nodes:
+            if node.type == 'OUTPUT_MATERIAL' and len(node.inputs[0].links) == 1:
+                outputNode = node
+                # Blender uses the last OUTPUT_MATERIAL node, so don't stop search here
```

Comment:
What is the order of output nodes when traversing the tree? is it guaranteed to be as we expect it, or it could change in the future?

Reply:
From my testing this is what blender does. The node order in the nodetree can change but blender always uses the last material output node. It might change but there is no other way for us to detect which material output node is used, it's undefined behaviour and up to the user to set up correctly.